design/baremetal-operator/user-defined-parameters.md

```diff
@@ -0,0 +1,456 @@
+
+### Goals
+
+- Add an optional field `extraParametersName` in BMH to allow specify a
+  `Secret` object.
+- Users can define arbitrary parameters in the `Secret`.
+- The parameters can be passed to ironic's driver_info.
+
+### Non-Goals
+
+- User-defined parameters can override the BMC parameters that set by existing
+  APIs.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I would like to build an OpenShift cluster with
+FIPS mode using Fujitsu PRIMERGY servers, but due to [the limitation of iRMC driver](https://github.com/openstack/ironic/blob/21.3.0/ironic/drivers/modules/irmc/common.py#L254-L257),
+I need to set `irmc_snmp_version` to "v3" and provide related credential.
+
+#### Story 2
+
+As a cluster administrator, when deploying Fujitsu PRIMERGY servers with
+firmware version >= `iRMC S6 2.00` using the iRMC driver, I need to provide
+additional parameters due to the following [change](https://docs.openstack.org/ironic/latest/admin/drivers/irmc#node-configuration):
+
+> Fujitsu server equipped with iRMC S6 2.00 or later version of firmware disables IPMI over LAN by default.
+> However user may be able to enable IPMI via BMC settings. To handle this change, irmc hardware type first
+> tries IPMI and, if IPMI operation fails, irmc hardware type uses Redfish API of Fujitsu server to provide
+> Ironic functionalities. So if user deploys Fujitsu server with iRMC S6 2.00 or later, user needs to set
+> Redfish related parameters in driver_info.
+
+Note that so called "Redfish related parameters" above is for iRMC rather than Redfish driver.
+
+## Design Details
+
+### Implementation Details/Notes/Constraints
+
+#### Add New Field
+
+Add a new string type field in `spec.bmc.extraParametersName` for `BareMetalHost`
+CRD to specify a `Secret` containing user-defined parameters.
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: worker-test-bmc-secret
+type: Opaque
+data:
+  username: YWRtaW4=                                      # admin
+  password: YWRtaW4=                                      # admin
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: worker-test-bmc-extra-parameters                  # the secret containing user-defined parameter
+type: Opaque
+data:
+  irmc_snmp_auth_password: YXNkZjEyMzQ=                   # asdf1234
+  irmc_snmp_priv_password: YXNkZjEyMzQ=                   # asdf1234
+stringData:
+  snmp_version: v3
+  irmc_snmp_user: testuser
+---
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-test
+spec:
+  online: true
+  bootMACAddress: <mac-address>
+  bmc:
+    address: <bmc-address>
+    credentialsName: worker-test-bmc-secret
+    extraParametersName: worker-test-bmc-extra-parameters # New field
+```
+
+#### Convert ExtraParametersName to ExtraParameters
+
+Get the specified `Secret` according to `ExtraParametersName`, and then convert
+the data of the `Secret` to `ExtraParameters`.
+
+```diff
```

Comment:
Please don't paste actual patches to specifications, it's not the place to review them. We need to agree on the design first.